### DISCUSSION TITLE
How to integrate own library with WebPack which needs to span WebWorkers and AudioWorklets

### DISCUSSION THREAD

**Discussion by Danielku15** [webpack/webpack#15382] [Q&A] [answered]
I am cross posting my StackOverflow question from [here](
https://stackoverflow.com/questions/70733934/how-to-integrate-own-library-with-webpack-which-needs-to-span-webworkers-and-aud), as it rather stays silent over there and I still hope to find a good solution to my problem. 

You can find the overall details on my question at SO but in very short to get you teased: I need to solve two key challenges with WebPack but feel rather stuck: 

1. How can I obtain the absolute URL to the chunk.js my library is put into so that I can spawn new Web Workers on this file? (red part)
2. How can I configure WebPack to call a certain entry-point within my library to run the Worker initialization logic when spawned using the URL from Q1? (orange part)

![image](https://user-images.githubusercontent.com/674916/153768471-dbca219a-d572-48d1-99e5-e933484b5dfb.png)

My library is rather agnostic on how it is used (UMD, ESM, Bundlers) and would prefer to keep it like that. I want to minimize the effort for the developers using my lib even when bundling it through WebPack into their apps.  

I'm looking forward to learn about the possibilities WebPack might offer to achieve this setup.

**Comment by alexander-akait** [comment]
Do you want to load the same chunk but in worker context (i.e. `chunk-my-lib.js`)? Do you looking this https://github.com/webpack/webpack/issues/12638?

**Reply by Danielku15** [reply]
> Do you want to load the same chunk but in worker context (i.e. chunk-my-lib.js)? 

In best case: Yes the worker should load the same chunk as the main browser thread (as my diagram indicates). This would ensure that:
 
1. The browser will not need to load an additional separate chunk or file but take the one already in the local browser cache. 
2. Any additional customizations (e.g. plugins) packed together into the chunk of my library will also be available within the worker.

> Do you looking this https://github.com/webpack/webpack/issues/12638?

https://github.com/webpack/webpack/issues/12638 sounds similar from the use case to my first problem I noticed but also the discussion there confuses me a bit. The problem as described in the issue is, that something like `'explicit-own-filename.js'` (`chunk-mylib.js` in my example) cannot be obtained. I couldn't find any code construct allowing me to access this filename after bundling. My attempts are documented within the SO post.

**Reply by alexander-akait** [reply]
Why you need URL to script? It is real bad idea, you should rely `import.meta.url`, name can be renamed/hashed/moved to another dir by bundlers

**Reply by Danielku15** [reply]
>Why you need URL to script?

Please read through my post on StackOverflow, I think it really describes well in full extend my concrete use case why I need the URL to the chunk. It is linked above in my initial post: https://stackoverflow.com/questions/70733934/how-to-integrate-own-library-with-webpack-which-needs-to-span-webworkers-and-aud

> you should rely import.meta.url

This does not work. `import.meta.url` is rewritten to a file URL pointing to the previous source of the module (before packing) by WebPack. This URL is pretty useless outside a local Node.js environment.  

> name can be renamed/hashed/moved to another dir by bundlers

This is exactly my use case. Bundlers (WebPack in this case) is taking my library and packing it into a chunk. I need to know the exact URL to this chunk within my library which was packed into this chunk. I don't need any access to my original file which was packed into the chunk.

**Reply by Danielku15** [reply]
I prepared you quickly a Stackblitz example: https://stackblitz.com/edit/github-hhejac?file=src/mylib.js

The `/dist/no-webpack.html` version as ES Module works fine. We can see in correct order: 
1. Hello Main Browser UI Thread!
2. Starting library
3. Output of all variants where `import.meta.url` points to the **url** providing mylib
4. An error from the `new Worker(import.meta.url)` which can be ignored. 
5. Hello Web Worker!

The `/dist/index.html` shows the WebPack bundled version. As you can see, none of the `getMyLibScriptFile` variants allow me to resolve the `chunk-mylib.js` within the now embedded `mylib.js`. Also looking at the rewritten source of `chunk-mylib.js` in getMyLibScriptFile we can see that no variant works well. 

1. `currentScript` is pointing to main.js
2. `__webpack__public_path__` points to `https://github-hhejac--5000.local.webcontainer.io/dist/`
3. `import.meta.url` was rewritten to the file-url where originally mylib.js was: `file:///home/projects/github-hhejac/src/mylib.js`
4. `new Url(import.meta.url, import.meta.url)` points to an additional JS asset file being generated with a hash. In my real project there is also additional code added to this file which makes it unusable in the web. 
5. `new Worker(import.meta.url)` is also rewritten to the file url. 

I need some way to resolve `/dist/chunk-mylib.js` to spawn my Worker.

**Reply by vankop** [reply]
`new Worker(import.meta.url)` does not work in webpack. you can use only
```
new Worker(new URL('./worker.js'), import.meta.url)
```

**Reply by Danielku15** [reply]
`new Worker(import.meta.url)` was "just" one of the many attempts to get one version which gives me the right urls. I would be happy to use something like `new Worker(new URL('./worker.js'), import.meta.url)` but this just brings me to the first problem again: I do not know what to write for `./worker.js` within `mylib` as it will be the chunk name which WebPack defines. 

`mylib` is usually an external library which comes through NPM and is then embedded by developers through WebPack into the final app. Hence I need a construct which I can put into my JavaScript which will result in correct behavior in WebPack. 

To go more towards a solution driven approach instead of discussing what does not work today: 

It would be great if there is something like `__webpack_chunk_name__ `, `import.meta.chunkUrl` or any other construct which will be replaced by WebPack with the absolute URL to `/dist/chunk-mylib.js`. Just the same way it already replaces `import.meta.url` with the file path. 

I see multiple possibilities where some might have unexpected side effects: 

- Introduce an officially available variable like `__webpack_chunk_name__` or `import.meta.chunkUrl` which will give you the URL of the chunk. 
- Introduce a new code construct which is detected and rewritten like `new URL(import.meta.url)` rather pointing to the chunk URL (which must not be a file:// url). In best case the construct would be compatible with all environments (web, node.js,..)
- Introduce a new setting into WebPack which changes the behavior of rewriting `import.meta.url`. It would allow you to configure WebPack to either: point to the file URL of the original source file (current behavior), point to the file URL of the chunk (new behavior 1 for node.js environments), point to the relative output path of the chunk respecting `__webpack_public_path__`.

**Reply by vankop** [reply]
then you need to bundle worker part separately..

**Reply by Danielku15** [reply]
Unfortunately uch one-sentence answers do not really help me any further. It does not provide me any real step forward how to really solve my case. It is impossible for me, with not being a WebPack expert, to derive what you actually propose to change within the setup of my library, and what I should advise to integrators to reach a working solution. 😞 I am hoping for a bit more constructive feedback which actually applies to my actual usecase and setup.

Workers have always been a huge pain in bundlers and I know that it is not a "simple" topic and might not apply to the majority of users which don't use any Web Worker or Audio Worklet features in their applications. But I still think it is a valid "advanced" topic which should be solvable in WebPack in some way. I as library author only have limited capabilities to suggest what actual integrators using WebPack, Rollup or Parcel can do. 😢 

To challenge your statement that I should change something: For me personally this behavior is a potentially missing feature or bug in WebPack. It should never ever put file urls to a source file into a packed output file. IMO in the context of bundling you would never expect someone to still access the original source files anymore but `import.meta.url` is translated to that. The only use-case I potentially see, might be debug related code like logging where you're interested into the original source, but for this sourcemaps exist. Also the file URLs will not work in any Web Context and considering that WebPack is the main bundler used in Angular, Vue and React enviroments this seems an incorrect behavior. 🤔

**Reply by alexander-akait** [reply]
Perhaps because the approach is not statically analyzed? Can you create github repository with library and application, create a simple readme what you have and what you expected to see (maybe small picture as you provide above). Again you should never rely on chunk filename, it is really bad idea, bundler can apply any count of optimizations and split chunks on two/three/more and more and more.

I see you use `new URL(import.meta.url, import.meta.url)`, that is interesting, what you expected? You can keep `new URL(import.meta.url, import.meta.url)` dynamically https://github.com/webpack/webpack/pull/15246, not released yet, but will be available in the next release

**Reply by Danielku15** [reply]
> Can you create github repository with library and application, create a simple readme what you have and what you expected to see (maybe small picture as you provide above). 

I created a example repo here: https://github.com/Danielku15/webpack-worker
Just start the repository with `npm i && npm start` and then open the website. You will find a real world example (with&without webpack) and a simplified setup with a small fake library trying to do the same thing: Spawning a worker using the file it is contained in. 

The samples without WebPack are showing what I want to achieve from a runtime behavior. The ones with WebPack show the challenge. In the `getMyLibScriptFile` of `mylib.js` I need to obtain somehow the single or multiple URLs required to startup correctly my library. 

 >Again you should never rely on chunk filename, it is really bad idea, bundler can apply any count of optimizations and split chunks on two/three/more and more and more.

I don't see a problem in this if WebPack allows me to obtain the correct URLs so that I can include/import them in the Workers. I cannot workaround the standard here. Workers (and Audio Worklets) simply need one file which acts as entry point for the background process. 
Also I don't see it as a problem if I advise the user to follow a specific chunk configuration for my library. I can live with some trade-offs as long the solution works and is simple for my integrators. 

> I see you use new URL(import.meta.url, import.meta.url), that is interesting, what you expected?

It was simply one of the many attempts to get somehow the URL to the chunk. I [read](https://webpack.js.org/guides/web-workers/) about the details that WebPack treats `new Url('./something.js', import.meta.url)` in a special way so I hoped to resolve the own file url with this mechanism.

**Reply by alexander-akait** [reply]
> I don't see a problem in this if WebPack allows me to obtain the correct URLs so that I can include/import them in the Workers. I cannot workaround the standard here. Workers (and Audio Worklets) simply need one file which acts as entry point for the background process.
> Also I don't see it as a problem if I advise the user to follow a specific chunk configuration for my library. I can live with some trade-offs as long the solution works and is simple for my integrators.

Why we need it in webpack, why we need browser API on webpack side? Why don't use `currentScript`

Also look at your code, you have `script` type and wait `import.meta.url` will work (we don't polyfill your code)? You can install the latest version of `webpack` and use:
```
module: {
        parser: {
            javascript: {
                importMeta: false
            }
        }
    },
```

It will keep `import.meta.url` as is, but you need to switch to ESM, otherwise `currentScript` is workaround, alternative - you can implement own plugin for webpack, but I don't think you are looking for this.

Here my example of polyfill:
```
const import_meta_url =
  typeof document === 'undefined' ? new URL('file:' + __filename).href :
    (document.currentScript && document.currentScript.src || new URL('main.js', document.baseURI).href);
```

**Reply by Danielku15** [reply]
> Why we need it in webpack, why we need browser API on webpack side? 

I'm not sure if I can follow you here 😅 (might be a bit of a language barrier here) . Let's see if I understand you right: you are wondering why webpack is supposed to offer a certain functionality which is supposed to be available through standard browser APIs? 

In this case my answer would be: With the latest PR merged above, I would have to check if `import.meta.url` is now working correctly and allows me to resolve the "quasi" entry point to my library. But generally I would also say: WebPack rewrites code quite significantly and is introducing an own custom module system. On top, as you indicated: WebPack does quite some complex parts with splitting potentially files apart into multiple chunks. 

Looking at the compile output of "chunk-mylib.js" or "chunk-alphatab.js": There is no standard ESM (import/export) in place anymore, the whole dependency management is done through WebPack. 

That's why I think only WebPack can offer the required features to provide to the rewritten/bundled libraries the URL of the file it is contained in. `import.meta.url` is rewritten today, and might not be rewritten tomorrow. But also with WebPack rewriting ESM to WebPack-require I would be surprised if the default browser API `import.meta.url` would provide me the URL to the file which contains my library. 

In fact, likely these chunks even are not usable by anyone outside the main WebPack entry point. 

While the technical solution is still unclear at this point. I hope the use case what I want to achieve is understandable.I can try to rephrase it another time if this helps in understanding what I want to achieve. ☺️ 

> Why don't use currentScript

Due to the fact how WebPack might resolve/initialize dependencies (e.g. by using eval in some cases) there is no guarantee that currentScript will point to the file which can be launched as worker. I think almost always this is the URL pointing to the main entry point of the application (not the library). 

> Also look at your code, you have script type and wait import.meta.url will work (we don't polyfill your code)? You can install the latest version of webpack and use:

I will give the setting a try to see how it behaves. But it is also a tricky situation: I am the author of `mylib` (actually `@coderline/alphatab`) which comes through NPM. My library is rather independent from the environment it will be used in. It ships an UMD (.js) and an ESM (.mjs) flavor. The ESM version could skip the `document.currentScript` but it's just there from the common typescript codebase. 

It's hard to convince all the integrators of my library to use a pre-release webpack to use my library in their projects. 

Also the polyfill approach would be maybe working if I am the author of the application using WebPack. I could write some wrappers, polyfills etc. which point to the correct location (even hard code it if required). But this is not a feasable solution for the market.

**Comment by Danielku15** [comment]
I think I get closer to an understanding on how WebPack works and I might be able to ask now differently in regards to my needs that would help me propose a solution to my users. Let's ignore for now the previous discussion we had and focus on this new setup. 

Assume following setup: We have an application with two entry points: `app.js` and `worker.js` where both might import various dependencies which might be put into chunks by WebPack. We want to launch from `app.js` a new WebWorker using `worker.js` as entry point using inline workers. Inline workers are important for CORS purposes as WebWorkers have quite strict rules. 

Here live examples.

## Using import statements in worker.js 
Source Code: https://github.com/danielku15/webpack-worker/tree/import-statement

![image](https://user-images.githubusercontent.com/674916/154815816-31eba7d1-dbd6-4f16-87cf-925e55a4db7f.png)

The html-webpack-plugin ensures that all JS files are included as script tag to the index.html and therefore app.js can start. 
But when we start a new WebWorker using inline workers with `importScripts` we have no possibility in app.js to generate `importScripts` statements for all required files. Also WebPack itself does not take any care to load all the chunks and dependencies the worker.js might have. Therefore `worker.js` will never be actually started but is stuck in a wait state forever because "somebody" will ever provide the dependencies. How could this be solved?

## Using import functions in worker.js
Source Code: https://github.com/danielku15/webpack-worker/tree/import-function
![image](https://user-images.githubusercontent.com/674916/154816124-89a9a97e-2b7d-41fc-9181-ab2d958d0d91.png)

When using `import('./mylib.js').then(...)` WebPack generates some different code. It tries to generate a `script` tag on the page to automatically load the dependency. But generating script tags in web workers will fail for obvious reasons. The import function seems not to be supported in WebWorkers, it assumes a browser UI context with a DOM available. 

Side note: I would still have to test how this would behave if `mylib.js` has further dependencies which need to be loaded, or if `mylib.js` is split into multiple chunks. But for now let's keep it simple.

**Comment by Danielku15** [comment, marked as answer]
I started now developing an own WebPack plugin and learned quite a lot on how some internals of WebPack work. While looking at the built-in plugins and hidden magic functions of WebPack I was able to identify all the bits I need to get a correct setup as I need. My intermediate solution is: 

1. In my library I added a detection for WebPack like: `typeof __webpack_require__ !== 'function'` to spawn workers compatible with a WebPack setup. There might be better/different ways to detect whether my library is embedded into WebPack and maybe I'll still change this. 
2. I trigger WebPack to generate a WebWorker wrapper code by spawning my workers with: `new Worker(new URL('@coderline/alphatab', import.meta.url))` where `@coderline/alphatab` is my library that can be resolved by WebPack. This wrapper code contains the WebPack bootstrapping, and even loads any dependency chunks using `importScripts`. This currently leads to a "0.js" being generated while I guess it should be a proper hash code, might report a more dedicated bug for that. 

The only problem with this solution is: CORS in cases CDNs are used. I am closing this question in favor of the discussion over at https://github.com/webpack/webpack/discussions/14066 which is about the support of inline Workers.

**Reply by Danielku15** [reply]
@alexander-akait Unfortunately it is hard to search for issues in GitHub related to a "0.js" output. Are you aware of any existing issue that WebPack generates a "0.js" file name for the WebWorker wrapper in case you specify a whole library like described above? Otherwise I would enter a new one.

**Comment by nthorn-adam** [comment]
I just faced a decently similar issue to this.

I am producing a React component library which is bundled / published with Rollup, but developed on top of and consumed downstream by react-scripts. Therefore I wanted this to be zero-config for react-scripts (Webpack 5), with the Worker published to do smart edge routing (expensive for main thread).

This guide is actually pretty helpful, but obviously doesn't explicitly address this case: https://webpack.js.org/guides/web-workers/

My solution had a few parts:
* First, I had to compile my worker to a separate file in the publish. I added an extra rollup config for just the worker.
* Second, based on NODE_ENV, I would either use the relative path to the worker in DEV, which react-scripts/Webpack would be able to resolve cleanly:
  * DEV: `new Worker(new URL('./worker.ts', import.meta.url))`
    * The worker is inside the component folder; it uses relative-only imports and avoids barrel modules. Webpack knows how to make this into a worker in dev, Rollup required a small amount of work.
  * PROD/publish: `new Worker(new URL(`${process.env.EDGEROUTER_RELATIVE_PATH_FROM_BUNDLE}`, import.meta.url))`
    * I inject this variable during publish, and it ends up looking like `./workers/edge-router.worker.js`.
    * In the outputted bundle you get in node_modules, for instance, the file `edge-router.worker.js` is a zero-dependency file (everything inlined), and the module consuming it loads the worker via `new Worker(new URL('./workers/edge-router.worker.js', import.meta.url))`
  * Pay careful attention to this line in that guide ^: `"Warning Using a variable in the Worker constructor is not supported by webpack ...  This is because webpack cannot analyse the syntax statically"`
  * So, you have to make sure that you do the injection like I did above directly into the `new URL` call, not a variable
* Finally you can just use the library downstream! Decently graceful, just very opaque while stumbling around.
   
Some caveats:
* This worker is actually used by one of the components in the library. I imagine you could summon it yourself with `new Worker(new URL(`@my/component-library/dist/workers/edge-router.worker.js`, import.meta.url))`, but I haven't tried of course
* This approach is probably coupled to Webpack/react-scripts 5, but I imagine with some custom config you could use it in other contexts

Not sure if this has value for you, but it was interesting for me. Performs beautifully too.
